design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+
+## Security Consideration
+
+### "auth_req_id" Swapping
+
+"auth_req_id" Swapping is that keycloak sends tokens to the entity other than the one from which keycloak received the backchannel authentication request.
+
+To prevent it, keycloak checks and confirms that CD sending Backchannel Auth Request bound with "auth_req_id" must be equal to CD sending Token Request with the same "auth_req_id". 
+
+### User Swapping
+
+User Swapping is that keycloak sends tokens generated based on the result of the authenticate and be authorized by the end-user whom CD did not intended.
+
+To prevent it, keycloak checks and confirms that the user to be authenticated must be equal to actually authenticated user. 
+
+### Replay Attack/Guessing Attack
+
+"auth_req_id" might be the replay attack and guessing attack target. To prevent this replay attack, the following measure are considered :
+
+* "auth_req_id" is short lived
+* "auth_req_id" is one time use (except for the token request before completion of AuthN/AuthZ by AD)
+* The generator of "auth_req_id" has enough bits of entropy to guard against these attacks in a certain level(Use a version 4 UUID string having about 112 bits of entropy)
+
+### DoS Attack
+
+ keycloak only aims to minimize CPU usage by Token Request aiming DoS attack.
+ keycloak cannot refuse to receive Token Request aiming DoS attack. If want to prevent it, other means is required.
+
+## Privacy Consideration
+
+### PII on the wire
+
+The user information might be conveyed by "login_hint", "login_hint_token", and "id_token_hint".
+It might be better that such the information be encoded so that other entity cannot identify the user itself. AuthN/AuthZ User Resolver mentioned just before can be used to convert this encoded information to User Model.
+
+## Performance Consideration
+
+### Sync Auth Req Context Cache and Auth Result Cache
+
+keycloak needs to share or distribute among nodes cache entries of Auth Req Context Cache and Auth Result Cache. Therefore, the performance problem might arise if keycloak has vast number of Backchannel Authentication Requests simultaneously or in the short time frame.
+
+### Find User Session Model Bound with "auth_req_id"
+
+In Token Req/Res part, keycloak needs to find User Session Model of the end-user authenticated by AD in a CIBA flow with bound with "auth_req_id".
+
+The one way is that keycloak sets "auth_req_id" to User Session's note when generating this User Session and find it afterwards by linear search. It seems to have serious performance problem especially when keycloak having vast number of User Sessions and distributed among several nodes.
+
+The faster way is using its User Session ID as a search key. Therefore, in Backchannel Auth Req/Res part, keycloak issues the User Session ID in advance and use it when generating User Session in AuthN/AuthZ by AD part. However, this way cannot utilize key affinity feature of Infinispan.
```

Comment:
In regards to authentication sessions vs cross-dc. I thought we don't push too much to our cross-dc support?

You won't need sticky sessions because the `auth_req_id` already references the session we need to lookup. It is just a matter of having the session shared across nodes. But if cross-dc support is a must, yeah, I think authentication sessions are not an option.

But I was thinking ... Once the user has consented to the client request (at the AD) we are going to create a session, right? Maybe, until that happens we don't need to store anything as everything we need to know is in the `auth_req_id`. Once the user consented and the user session is created, the decision to whether or not the token request should be fulfilled can be based on a simple query in the "sessions" cache. Makes sense? Is that what you are proposing?

Reply:
I assume that cross-dc support will be good. IMO the fact that it is "Technology preview" doesn't mean that we can relax on supporting it for some Keycloak features. But maybe I am wrong here.

Per https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#rfc.section.10.1.1 I see this:  Once redeemed for a successful token response, the auth_req_id value that was used is no longer valid.

So this probablb means that it is supposed to be single-use?

Reply:
The single-use behavior I meant is in regards to using the CodeToTokenStore.

We still need the auth_req_id to be single-use, but we can do it like that:

1) Client sends the authentication request
2) Server generates an auth_req_id using a JWT format containing some identifier that we can use later to pass to the AD (so we can recognize that a request from AD is related to the client authentication request). This JWT may have some expiration, iat, azp, etc, to make sure we only accept auth_req_id from clients within a period.
3) The server sends the request to AD to authenticate the user
4) AD ask user to approve the solicited authentication request. Validates the binding_message if necessary, etc.
5) Once the user approved the request, AD sends back the auth_req_id
6) The server creates a user session based on the auth_req_id

Clients pooling the server for a token will send the auth_req_id and once a user session exists and is associated with the auth_req_id, we return the tokens. Otherwise, the server responds with a message whenever the auth_req_id expires and is no longer invalid. If the client tries to use the token again, after a determined period, it won't be possible.

We don't necessarily need to pass the auth_req_id to the AD. But only some reference that we can associate with the auth_req_id. Some data we put inside the auth_req_id token. But I'm not sure yet how bad is to just pass the auth_req_id to AD, considering the AD would not be able to obtain tokens (the server requires client authentication) and the token would not leak if using a secure protocol.

Reply:
> Clients pooling the server for a token will send the auth_req_id and once a user session exists and is associated with the auth_req_id, we return the tokens. Otherwise, the server responds with a message whenever the auth_req_id expires and is no longer invalid. If the client tries to use the token again, after a determined period, it won't be possible. When processing the auth_req_id for the first time after user approval, the session will not be tied with the auth_req_id yet, so we know it is the first time the auth_req_id is used.

Yes, the CodeToTokenStore provider (or SingleUseTokenStoreProvider) is useful in the cases when you want to guarantee single-use across whole cluster/cross-dc environment. Which is maybe not a strict requirement in the CIBA specification (unlike OIDC "code" parameter where the strict requirement is). If you want to guarantee strict single-use, then you may always end with the scenario like this:
a) Client sends the tokenRequest, which will be processed on Keycloak node1
b) Keycloak node1 will see that userSession is not yet associated with auth_req_id
c) At the same time, client sends the same tokenRequest, which will be processed on Keycloak node2
d) Keycloak node2 will see that userSession is not yet associated with auth_req_id
e) In the meantime request1 will associate the "auth_req_id" with userSession
f) At the same time, concurrent request2 will also associate the "auth_req_id" with userSession
g) both requests will successfully return TokenResponse. Which is bad if we want to strictly guarantee single-use

I don't know if this would be a problem? If not, then we don't need strict single-use and the CodeToTokenStore/SingleUseTokenStoreProvider at all.

The (a) and (c) needed to end of different cluster node, assuming that we will do "throttling", which will guarantee the KEycloak reject concurrent requests from client sent to same cluster node as discussed in other sub-thread :)
 
> We don't necessarily need to pass the auth_req_id to the AD. But only some reference that we can associate with the auth_req_id. Some data we put inside the auth_req_id token. But I'm not sure yet how bad is to just pass the auth_req_id to AD, considering the AD would not be able to obtain tokens (the server requires client authentication) and the token would not leak if using a secure protocol.

Yes, I believe the "Auth Result ID", which @tnorimat proposed in the design is the reference, which can be eventually saved as one of the claims in the "auth_req_id" JWT/JWE? So it looks that we're on same page here. I am also not sure how bad it is if "auth_req_id" is shared with the AD. I assume this can be bad based on what @tnorimat mentioned.

Reply:
It would be a problem. Considering user sessions are cross-dc aware, doesn't ISPN provide the necessary means to avoid this scenario?

Reply:
@pedroigor You mean the scenario when you have 2 concurrent requests to update same userSession and hence you expect one of them to fail? Yes, there are ways to avoid that. For example there is 3 arguments "cache.replace" method (The item is replaced just if the current value is equal to the new value). However, I recall some issues with that few years ago when worked on cross-dc stuff. So the "guaranteed" way is the single-use cache (and CodeToTokenStoreProvider), which should reliably ensure the single-use. This stuff is used for example for OIDC codes and IMO it works fine and would work fine here as well.

Reply:
Yeah, something like that. I'm 100% for your proposal to use the `CodToTokenStoreProvider`. The only concern I have in this regard is that cross-dc usually requires additional steps that might not be necessary when in a single cluster scenario.

So any overhead we add in order to support cross-dc affect people doing single-dc. IMO, we should also try to be as stateless as possible. All that makes a huge difference when deploying and scaling Keycloak, especially when doing single-dc which I think is the majority of use cases?

Reply:
Well, even in single-dc you can have the issue with 2 concurrent threads trying to update same entity. And for example, before we added cross-dc support, we had the issue with single-use code, which was not really single use (even in single-dc it could happen that same "code" was used multiple times). Regarding implementation, the "CodeToTokenStoreProvider" or "SingleUseProvider" should guarantee in both single-dc and cross-dc that particular stuff is really single-used, so there is no additional overhead in the "consumer" code.

But of course, in cross-dc environment, there is additional overhead due the network latency etc. So there is always some price to pay for cross-dc support. But people, who want to use cross-dc, should expect this additional price to pay.

Reply:
@mposolda 
>I can imagine that UserSessionID is even the same as "Auth Result ID", but I don't know if it has security implications? As I guess that the "Decoupled Authentication server" should not be able to know the UserSessionID, however he will be able to know "Auth Result ID" ?

As you suggested before, "Auth Result ID" is retained in `actionToken` cache by using `CodeToTokenStoreProvider`  or refactored one, I think that UserSessionID can be stored on this cache entry so that we do not need to send UserSessionID to "Decoupled Authentication Server". WDYT?

Reply:
@tnorimat +1